pkg/network/port.go

```diff
@@ -1,54 +1,89 @@
+// +build linux
+
 package network
 
 import (
+	"fmt"
 	"path"
 	"sync"
 	"time"
 
+	"github.com/DataDog/datadog-agent/pkg/process/util"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
 // PortMapping tracks which ports a pid is listening on
 type PortMapping struct {
-	procRoot    string
-	collectTCP  bool
-	collectIPv6 bool
-	ports       map[uint16]struct{}
+	procRoot     string
+	collectTCP   bool
+	collectIPv6  bool
+	ports        map[string]interface{}
+	defaultNetNs uint64
 	sync.RWMutex
 }
 
 //NewPortMapping creates a new PortMapping instance
-func NewPortMapping(procRoot string, collectTCP, collectIPv6 bool) *PortMapping {
+func NewPortMapping(procRoot string, collectTCP, collectIPv6 bool, defaultNetNs uint64) *PortMapping {
 	return &PortMapping{
-		procRoot:    procRoot,
-		collectTCP:  collectTCP,
-		collectIPv6: collectIPv6,
-		ports:       make(map[uint16]struct{}),
+		procRoot:     procRoot,
+		collectTCP:   collectTCP,
+		collectIPv6:  collectIPv6,
+		ports:        make(map[string]interface{}),
+		defaultNetNs: defaultNetNs,
 	}
 }
 
-// AddMapping indicates that something is listening on the provided port
+// AddMapping adds/overwrites a port mapping in the default network
+// namespace
 func (pm *PortMapping) AddMapping(port uint16) {
 	pm.Lock()
 	defer pm.Unlock()
 
-	pm.ports[port] = struct{}{}
+	pm.ports[portMappingKey(pm.defaultNetNs, port)] = struct{}{}
 }
 
-// RemoveMapping indicates that the provided port is no longer being listened on
+// RemoveMapping removes a port mapping in the default network
+// namespace
 func (pm *PortMapping) RemoveMapping(port uint16) {
 	pm.Lock()
 	defer pm.Unlock()
 
-	delete(pm.ports, port)
+	delete(pm.ports, portMappingKey(pm.defaultNetNs, port))
+}
+
+// AddMappingWithNs adds a port mapping in the given network namespace
+func (pm *PortMapping) AddMappingWithNs(nsIno uint64, port uint16) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	pm.ports[portMappingKey(nsIno, port)] = struct{}{}
+}
+
+// RemoveMappingWithNs removes a port mapping in the given network namespace
+func (pm *PortMapping) RemoveMappingWithNs(nsIno uint64, port uint16) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	delete(pm.ports, portMappingKey(nsIno, port))
 }
```

Comment:
Do we need both AddMappingWithNs and AddMapping ? It looks like we're only calling AddMapping/RemoveMapping in tests, and all the non-test code uses the ns aware code (which it should).

Could we change AddMapping/RemoveMapping to take an ns inode